components/org.apache.stratos.rest.endpoint/src/main/java/org/apache/stratos/rest/endpoint/api/StratosApiV41Utils.java

```
@@ -980,50 +981,41 @@ public static void addServiceGroup(GroupBean serviceGroupDefinition) throws Rest
+            //validate the group definition to check if groups duplicate in any groups and
+            //validate the group definition to check for cyclic group behaviour
+            validateGroupDuplicationInGroupDefinition(serviceGroupDefinition, new ArrayList<String>());
 
-                CloudControllerServiceClient ccServiceClient = getCloudControllerServiceClient();
+            CloudControllerServiceClient ccServiceClient = getCloudControllerServiceClient();
 
-                cartridgeNames = new String[cartridgeTypes.size()];
-                int i = 0;
-                for (String cartridgeType : cartridgeTypes) {
-                    try {
-                        if (ccServiceClient.getCartridge(cartridgeType) == null) {
-                            // cartridge is not deployed, can't continue
-                            log.error("Invalid cartridge found in cartridge group " + cartridgeType);
-                            throw new RestAPIException("No Cartridge Definition found with type " + cartridgeType);
-                        } else {
-                            cartridgeNames[i] = cartridgeType;
-                            i++;
-                        }
-                    } catch (RemoteException e) {
-                        throw new RestAPIException(e);
-                    } catch (CloudControllerServiceCartridgeNotFoundExceptionException e) {
-                        throw new RestAPIException(e);
+            cartridgeNames = new String[cartridgeTypes.size()];
+            int j = 0;
+            for (String cartridgeType : cartridgeTypes) {
+                try {
+                    if (ccServiceClient.getCartridge(cartridgeType) == null) {
+                        // cartridge is not deployed, can't continue
+                        log.error("Invalid cartridge found in cartridge group " + cartridgeType);
+                        throw new RestAPIException("No Cartridge Definition found with type " + cartridgeType);
```

@lahirus @imesh  IMO I think we should use something like "InvalidCartridgeGroupDefinitionException" because there is nothing wrong with the cartridge and it is the group definition which is incorrect. 
And there are 3 invalid scenarios in group definition, 1. duplication of cartridges, 2. duplication of groups, 3. cyclic behavior of groups. 
Should I throw 3 different exceptions or throw something like "InvalidCartridgeGroupDefinitionException", WDYT?